src/dreamchecker/lib.rs

```diff
@@ -545,6 +545,91 @@ pub fn check_var_defs(objtree: &ObjectTree, context: &Context) {
 
 // ----------------------------------------------------------------------------
 // Procedure analyzer
+#[derive(Debug)]
+pub struct ControlFlow {
+    pub returns: bool,
+    pub continues: bool,
+    pub breaks: bool,
+    pub fuzzy: bool,
+}
+
+impl ControlFlow {
+    pub fn alltrue() -> ControlFlow {
+        ControlFlow {
+            returns: true,
+            continues: true,
+            breaks: true,
+            fuzzy: false,
+        }
+    }
+    pub fn allfalse() -> ControlFlow {
+        ControlFlow {
+            returns: false,
+            continues: false,
+            breaks: false,
+            fuzzy: false,
+        }
+    }
+    pub fn terminates(&self) -> bool {
+        return !self.fuzzy && ( self.returns || self.continues || self.breaks )
+    }
+
+    pub fn terminates_loop(&self) -> bool {
+        return !self.fuzzy && ( self.returns || self.breaks )
+    }
+
+    pub fn no_else(&mut self) {
+        self.returns = false;
+        self.continues = false;
+        self.breaks = false;
+    }
+
+    pub fn merge(&mut self, other: ControlFlow) {
+        if !self.fuzzy && other.returns {
+            self.returns = true;
+        }
+        if other.fuzzy {
+            self.returns = false;
```

Comment:
?

Reply:
loops basically

Reply:
you can't (yet) determine if a loop ran, so it's "fuzzy" if it can actually return and terminate the proc leaving the code below the loop unreachable.